vuvuzela_test.go

```diff
@@ -0,0 +1,304 @@
+	msg := new(convo.DeadDropMessage)
+	msg.DeadDrop = *h.agreedDeadDrop
+	rand.Read(msg.EncryptedMessage[:])
+	out = append(out, msg)
+	h.sentOutgoing <- roundAndDeadDropMsgs{round: round, messages: out}
+	return out
+}
+
+func (h *mockHandler) Replies(round uint32, replies [][]byte) {
+	h.sentReplies <- roundAndReplies{round: round, replies: replies}
+}
+func (h *mockHandler) NewConfig(chain []*config.SignedConfig) {
+	h.newConfig <- chain
+}
+func (h *mockHandler) Error(err error) {
+	log.Errorf(h.errPrefix+": client error: %s", err)
+}
+func (h *mockHandler) DebugError(err error) {
+	log.Errorf(h.errPrefix+": debug client error: %s", err)
+}
+func (h *mockHandler) GlobalAnnouncement(message string) {
+	log.Debugf("GLOBAL ANNOUNCEMENT: %s", message)
+}
+
+func TestVuvuzela(t *testing.T) {
+	u := createVuvuzelaUniverse()
+	defer func() {
+		// Give time for goroutines to finish before pulling the rug out from under them
+		time.Sleep(1 * time.Second)
+		u.Destroy()
+	}()
+
+	aliceClient := u.createUser("alice")
+	_, err := aliceClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Alice's conversation: %s", err)
+	}
+	defer aliceClient.CloseConvo()
+
+	bobClient := u.createUser("bob")
+	_, err = bobClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Bob's conversation: %s", err)
+	}
+	defer bobClient.CloseConvo()
+
+	aliceRoundAndDeadDropMsgs := <-aliceClient.Handler.(*mockHandler).sentOutgoing
+	aliceRound := aliceRoundAndDeadDropMsgs.round
+	aliceMsgs := aliceRoundAndDeadDropMsgs.messages
+
+	bobRoundAndDeadDropMsgs := <-bobClient.Handler.(*mockHandler).sentOutgoing
+	bobRound := bobRoundAndDeadDropMsgs.round
+	bobMsgs := bobRoundAndDeadDropMsgs.messages
+
+	aliceRoundAndReplies := <-aliceClient.Handler.(*mockHandler).sentReplies
+	aliceReplyRound := aliceRoundAndReplies.round
+	aliceReplies := aliceRoundAndReplies.replies
+
+	bobRoundAndReplies := <-bobClient.Handler.(*mockHandler).sentReplies
+	bobReplyRound := bobRoundAndReplies.round
+	bobReplies := bobRoundAndReplies.replies
+
+	if len(bobMsgs) != 1 {
+		t.Fatalf("bob should have only sent 1 message")
+	}
+	if len(bobMsgs) != len(aliceReplies) {
+		t.Fatalf("bob should have as many messages as alice sent")
+	}
+	if len(aliceMsgs) != 1 {
+		t.Fatalf("alice should have only sent 1 message")
+	}
+	if len(aliceMsgs) != len(bobReplies) {
+		t.Fatalf("alice should have as many messages as bob sent")
+	}
+
+	latestRound, err := aliceClient.LatestRound()
+	if err != nil {
+		t.Fatalf("Failed to get latest round: %s", err)
+	}
+	if bobRound > latestRound || aliceRound > latestRound || aliceReplyRound > latestRound || bobReplyRound > latestRound {
+		t.Fatalf("Latest round %d should be same or later than %d, %d %d, %d", latestRound, bobRound, aliceRound, aliceReplyRound, bobReplyRound)
+	}
+
+	prevConvoConfig := u.ConvoConfig
+	prevConvoConfigInner := prevConvoConfig.Inner.(*convo.ConvoConfig)
+	newConvoConfig := &config.SignedConfig{
```

Comment:
Maybe start with the previous config and just tweak the few fields that change? That should reduce verbosity.

Reply:
Wouldn't that be a shallow copy, though? It will fail with `new config not created after prev config` since they both have the same `Created` time.

Reply:
nvm, did what you did in another test with marshal and unmarshal